### PR TITLE
Model sync status reporting; logging fixes; refactoring

### PR DESCRIPTION
**Pull request by glennmatthews**
A bit of a grab-bag here, my apologies!

- `DiffSyncModel`s now have `set_status` and `get_status` APIs, with the idea being that a model's `create`/`update`/`delete` method can use `set_status` to report statuses and messages other than the default, generic `(SUCCESS, "Created successfully")`. 
  - This is sort of a sideways approach to #35, as while it doesn't provide any automatic validation of whether a sync actually updated the backend, an implementer can now at least start with boilerplate CRUD functions that do something like `set_status(FAILED, "backend sync not implemented yet")` as a reminder to the implementer/user.
- This necessitated updating our minimum Pydantic version from 1.6 to 1.7 to pick up support for `PrivateAttr` attributes on a DiffSyncModel.
- In updating the sync logging logic to account for a model's reported status, I discovered a bug -- if a CRUD method fails gracefully (i.e. returns `None` instead of throwing an `ObjectCrudException`), we would get a log message for the failure *but also* still get a "success" log for that object as well. I've added test coverage in `test_diffsync.py` to catch this bug and ones like it.
- The updated logging logic pushed `__init__.py` over the arbitrary 1000-line limit imposed by pylint, which was already also complaining about the complexity of the `_sync_from_diff_element` function and the number of arguments being passed around between private sync helper functions. I got carried away somewhat and:
  - Similar to what was done in #20 to extract diff calculation logic into a `DiffSyncDiffer` stateful helper class, I've extracted the sync logic out of `DiffSync` and into a stateful `DiffSyncSyncer` helper class. Having some internal state in `DiffSyncSyncer` allowed me to reduce the number of parameters being passed around between internal methods and generally make the logic a bit cleaner.
  - To address the 1000-line module complaint, I moved the enums and flags into a `diffsync.enum` submodule, and moved `DiffSyncDiffer` and `DiffSyncSyncer` into a `diffsync.workers` submodule.

### CODE REVIEW

**Comment by glennmatthews** [comment]
A minor bug was introduced by #40 - due to some short-circuit logic introduced there,  if there were no changes present in the entire diffset, then even if `LOG_UNCHANGED_RECORDS` was enabled, no logs would be generated for the unchanged records. The latest commit fixes this.

**Review by dgarros — APPROVED** [review]
LGTM minus the name of the file but it's a minor detail

File: diffsync/workers.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
+"""DiffSync worker classes for diff and sync operations.
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply diffsync/workers.py:1]: Those both seem too generic to me. Would `operator` or `operations` be acceptable?

[the remark this answers is not in the document]
Comment (glennmatthews) [reply diffsync/workers.py]: Renamed to `helpers.py` per our Slack discussion.

**Review by dgarros — APPROVED** [review]

**Pull request merged**